Analyzer & code fix for ~IDE0047~ FS3583: remove unnecessary parentheses

brianrourkeboll opened the pull request:
Resolves #16078.

I have an analyzer and code fix for [IDE0047: remove unnecessary parentheses](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0047-ide0048) for expressions and patterns (adding it for type signatures can be done but will require updating `SyntaxTraversal.Traverse` to traverse types in many more places).

https://github.com/dotnet/fsharp/assets/14795984/c613a3b1-8518-428a-9625-1497aadef025

I shoehorned the call to the analysis logic into `DocumentDiagnosticAnalyzer.fs`. This lets us avoid making any upstream changes to Roslyn that then need to flow back here, although eventually just hooking every new analyzer into `DocumentDiagnosticAnalyzer` seems like it would become unsustainable or at least undesirable, so we should probably come up with some kind of longer-term plan at some point.

I've added a fair number of obvious tests, but, due to the combinatorially-explosive nature of the problem at hand, they are by no means exhaustive. I'd be happy to hear suggestions for ways to improve them.

<details>
<summary>Top-level test summary</summary>

![image](https://github.com/dotnet/fsharp/assets/14795984/8d601b4c-0555-4e82-ae6f-65fba849f965)

</details>

### Questions

- [x] The fix-all/bulk-fix functionality doesn't seem to trigger. Does anyone see an obvious reason why it wouldn't?
  - Edit: see https://github.com/dotnet/fsharp/pull/16079#issuecomment-1772926733.
  - This has been addressed in dab4696 by using a new FSxxxx diagnostic ID, FS3583, instead of IDE0047.
- [x] I assume I need to add `IFSharpUnnecessaryParenthesesDiagnosticAnalyzer` and `FSharpIDEDiagnosticIds.RemoveUnnecessaryParentheses` to Roslyn. ~~Might that also be why the bulk-fix functionality is not working for this (because Roslyn is not aware that IDE0047 is supported for F#)?~~ What are the logistics for getting that into Roslyn so that it is available to use here? **Edit**: it sounds like we don't actually want to do this, at least right now, but I at least moved the diagnostic creation logic into its own file in a2c9fe2.
- [x] When merged, should the analyzer be enabled by default in VS? It is for C#/VB, and I think it should be for F#, too (assuming the perf is acceptable). Either way, should it have a VS setting/be toggleable? As of right now, it will be enabled by default and will not be toggleable.
- [x] The current implementation independently traverses the entire untyped syntax tree using a creative call to the code underlying `SyntaxTraversal.Traverse`. Is this OK, or even the right way or place at all? In theory the diagnostics could be emitted in `CheckExpressions.fs`, `CheckPatterns.fs`, etc., during the type-checking pass, although it's debatable whether that's the right place to emit an "IDE" diagnostic. Or is there another way to avoid/amortize/consolidate full AST traversals across analyzers (if that's even desirable—I can see not wanting one misbehaving analyzer to be able to block others)?
- [x] I made a couple small updates to `SyntaxTraversal.Traverse` to handle some (seemingly) obvious omissions, namely:
    - Traverse the full patterns in the `parsedData` of the first `SynExpr.Lambda` in a lambda sequence _instead of_ the simple pats of each lambda in the sequence.
    - Traverse the full body of the first `SynExpr.Lambda` in a lambda sequence _instead of_ the body of each lambda in the sequence.
    - Traverse a record pattern's field patterns.
    - Traverse a quote pattern's expression.
  
  Is this OK?
- [x] The only public API addition is
  ```fsharp
  val getUnnecessaryParentheses: getSourceLineStr: (int -> string) -> parsedInput: ParsedInput -> Async<range seq>
  ```
  
  Is this the signature we want?
  
  I can think of other scenarios where we might additionally want to expose another function along the lines of
  
  ```fsharp
  val parenthesesNeededBetween: getSourceLineStr: (int -> string) -> node: SyntaxNode -> path: SyntaxNode list -> bool
  ```
- [x] It's not currently possible to apply the code fix to types (signatures or annotations, e.g., `x : (int)` → `x : int`), since `SyntaxTraversal.Traverse` does not traverse most `SynType` uses (and it explicitly skips `SynType.Paren` nodes altogether anyway). While certainly less pressing than expressions and patterns, would it ever make sense to handle types? (Edit: this should probably be done in one or more separate PRs.)
- [ ] It would be nice if we could assert that the AST never changed (at all? meaningfully?) upon application of the code fix in response to generated inputs, perhaps using something like FsCheck (compare https://blog.emillon.org/posts/2020-08-03-fuzzing-ocamlformat-with-afl-and-crowbar.html). If we did this, we'd need to take care to avoid any potential flakiness (or not run it on every commit, etc.). We could alternatively at least make such an assertion for the existing (non-generated) tests.
- [x] I'm not currently handling the ML infix operators (`mod`, `land`…), since all ([except for `mod`](https://github.com/fsharp/fslang-design/blob/526b2040d9522fabff9b5e091d814fe6f2631519/FSharp-6.0/FS-1114-ml-compat-revisions.md#use-of-land-lor-lxor-lsl-lsr-asr)) now emit warnings. Should I? **Edit**: it seems reasonable not to handle them, since the default behavior will simply be not to suggest that parens can be removed even if they theoretically could, which I suspect is for the best, since I don't think many F# programmers without an OCaml background would know their precedence off the top of their head.

### When ready

- [x] Update surface area.

### Remaining work

- [x] Add tests that call the new FCS API directly.
- [x] Consolidate some framework-y test code.
- [ ] Clean up/organize tests.

### Potential followups

- Add fix-all tests.
- Add AST-checking tests(?). (I have a version that does this, and it does provide some amount of safety, but it surfaces a few mostly superficial representational annoyances that may or may not be worth spending the time to deal with.)

nojaf commented on src/Compiler/Service/ServiceParseTreeWalk.fs:
```diff
@@ -575,6 +582,14 @@ module SyntaxTraversal =
 
                     if ok.IsSome then ok else traverseSynExpr synExpr
 
+                | SynExpr.Lambda (inLambdaSeq = false; body = synExpr; parsedData = Some (pats, _)) ->
```
This feels weird to introduce without a code comment.
Why is the `bodyExpr` from `parsedData` not used?
The [inner lambda](https://fsprojects.github.io/fantomas-tools/#/ast?data=N4KABGBEAmCmBmBLAdrAzpAXFSAacUiaAYmolmPAIYA2as%2BEkAxgPZwWTwCuyYAHmACeYAF5gAtAD4wACgCUeApFj8ADlWTQKAFwBO3WCAC%2BQA) will be processed again, so I believe the pats will be visited multiple times here.

brianrourkeboll replied:
…But as for why I didn't add traversal of the `bodyExpr` of `parsedData`: it seems like it's the same as the outer `body` (both are just `SynExpr`s). Do you know whether there is any case where it isn't?

nojaf replied:
See https://fsprojects.github.io/fantomas-tools/#/ast?data=N4KABGBEAmCmBmBLAdrAzpAXFSAacUiaAYmolmPAIYA2as%2BEkAxgPZwWTwCuyYAHmACeYAF5gAtAD4wACgCUeApFj8ADlWTQKAFwBO3WCAC%2BQA

```fsharp
fun x y z -> ()
```

![image](https://github.com/dotnet/fsharp/assets/2621499/d3bb4e3c-76e9-4ce4-ab68-d858228da227)

brianrourkeboll replied:
Ah, yes. Does f7403f4 look better? I don't think we need to traverse `bodyExpr` since it will eventually be visited as part of `body`.

nojaf replied:
Hmm, I'm not overly sure but I think in general it might make more sense to just traverse `bodyExpr` and not bother with `body` in this case.

This question is larger than the scope of this PR. What was already in the visitor code was questionable to begin with.

The untyped AST for Lambdas is already optimized to work with during type-checking.

When a user has `fun x y z -> ()`, the AST is already expanded:

```fsharp
SynModuleDecl.Expr(
    expr =
        SynExpr.Lambda(
            body =
                SynExpr.Lambda(
                    body =
                        SynExpr.Lambda(
                            body = SynExpr.Const(constant = SynConst.Unit, range = R("(1,13--1,15)")),
                        ),
                ),
            parsedData =
                Some(
                    ([
                        SynPat.Named(
                            ident = SynIdent.SynIdent(ident = Ident("x", R("(1,4--1,5)")), trivia = None),
                            isThisVal = false,
                            accessibility = None,
                            range = R("(1,4--1,5)")
                        )
                        SynPat.Named(
                            ident = SynIdent.SynIdent(ident = Ident("y", R("(1,6--1,7)")), trivia = None),
                            isThisVal = false,
                            accessibility = None,
                            range = R("(1,6--1,7)")
                        )
                        SynPat.Named(
                            ident = SynIdent.SynIdent(ident = Ident("z", R("(1,8--1,9)")), trivia = None),
                            isThisVal = false,
                            accessibility = None,
                            range = R("(1,8--1,9)")
                        )
                     ],
                     SynExpr.Const(constant = SynConst.Unit, range = R("(1,13--1,15)")))
                ),
            range = R("(1,0--1,15)"),
            trivia = { ArrowRange = Some(R("(1,10--1,12)")) }
        ),
    range = R("(1,0--1,15)")
)
```

The user did not write `fun x -> fun y -> fun z`, so I wouldn't follow that path to begin with in the visitor code.

I cannot think of any scenarios where you would want to do this from a tooling perspective. Although maybe I lack imagination.

I'm not hellbent on this, this is just my take on it.
@baronfel what are your thoughts on this one?

brianrourkeboll replied:
8e493e1

brianrourkeboll commented:
@nojaf

> > Add a prefix: string option (or trivia) field to the SynConst cases for numeric literals to indicate the presence of prefixed +, -.
> 
> When it comes to strings and other constants, there actually are a lot of scenarios where the value in the untyped AST doesn't reflect what the user wrote. The way we deal with this in Fantomas is to get the original text from `ISourceText` (using #15979).

Hmm yes, I tried the following and it does work, although it pollutes the signature/callsite a bit for what seems like a rather small thing.

```fsi
val getUnnecessaryParentheses (getTextAtRange: range -> string) (parsedInput: ParsedInput) : Async<range seq>
```

```fsharp
// In the body of getUnnecessaryParentheses.
let (|Text|) = getTextAtRange
let (|StartsWith|) (s: string) = s[0]
let rec (|PrefixedNumericLiteral|_|) = function
    | SynExpr.Const (SynConst.SByte _, Text (StartsWith ('-' | '+')))
    | SynExpr.Const (SynConst.Int16 _, Text (StartsWith ('-' | '+')))
    | SynExpr.Const (SynConst.Int32 _, Text (StartsWith ('-' | '+')))
    | SynExpr.Const (SynConst.Int64 _, Text (StartsWith ('-' | '+')))
    | SynExpr.Const (SynConst.Byte _, Text (StartsWith ('-' | '+')))
    | SynExpr.Const (SynConst.UInt16 _, Text (StartsWith ('-' | '+')))
    | SynExpr.Const (SynConst.UInt32 _, Text (StartsWith ('-' | '+')))
    | SynExpr.Const (SynConst.UInt64 _, Text (StartsWith ('-' | '+')))
    | SynExpr.Const (SynConst.Decimal _, Text (StartsWith ('-' | '+')))
    | SynExpr.Const (SynConst.Double _, Text (StartsWith ('-' | '+')))
    | SynExpr.Const (SynConst.Single _, Text (StartsWith ('-' | '+')))
    | SynExpr.Const (SynConst.UserNum (StartsWith ('-' | '+'), _), _) -> Some PrefixedNumericLiteral
    | SynExpr.Const (SynConst.Measure (constant, m, _, _), _) -> (|PrefixedNumericLiteral|_|) (SynExpr.Const (constant, m))
    | _ -> None
```

```fsharp
let getTextAtRange m = sourceText.ToString(RoslynHelpers.FSharpRangeToTextSpan(sourceText, m))
let! unnecessaryParentheses = UnnecessaryParentheses.getUnnecessaryParentheses getTextAtRange parseResults.ParseTree
```

Do you think it's worth the addition? We could pass in the Roslyn `SourceText` or F# `ISourceText` instead, although we don't actually need any other functionality from them.

nojaf commented:
I think having it as `val getUnnecessaryParentheses: sourceText: ISourceText -> parsedInput: ParsedInput -> Async<range seq>` will be most convenient for most users. `Roslyn SourceText` will probably be too tied to the VS side of things. It would be a shame if FSAC can't benefit from this work.
 
Overall it is a trade-off for sure. Passing in `ISourceText` seems reasonable to me.

brianrourkeboll commented:
> I think having it as `val getUnnecessaryParentheses: sourceText: ISourceText -> parsedInput: ParsedInput -> Async<range seq>` will be most convenient for most users. `Roslyn SourceText` will probably be too tied to the VS side of things. It would be a shame if FSAC can't benefit from this work.
> 
> Overall it is a trade-off for sure. Passing in `ISourceText` seems reasonable to me.

Hmm, I'm not totally against using `ISourceText` instead of `getSubtext: (range -> string)`, but

1. I'm just noticing 🤦 that other public APIs in `ServiceAnalysis.fs` take `getSourceLineStr: (int -> string)` functions[^1]
2. It seems to me that the most compelling reasons to pass in an interface rather than a function are:
  a. If we foresee that the consuming function may someday need to use other APIs exposed by the interface
  b. If we want to do something that we can't use a first-class function for, like return a `ReadOnlySpan<char>` instead of a `string` to avoid allocations (and since we don't actually need a `string`)

As for (1), `getSourceLineStr: (int -> string)` would work fine, although the real requirement is the ability to match against an arbitrary span of subtext rather than anything to do with source lines.

As for (2a), I don't really foresee that this particular function (`getUnnecessaryParentheses`) will ever need to use any of the APIs that `ISourceText` currently exposes other than `GetLineString` (as a roundabout way to get a specific character or span of characters to compare against) or just `GetSubTextFromRange` (after #15979):

```fsharp
type ISourceText =
    abstract Item: index: int -> char with get
    abstract GetLineString: lineIndex: int -> string
    abstract GetLineCount: unit -> int
    abstract GetLastCharacterPosition: unit -> int * int
    abstract GetSubTextString: start: int * length: int -> string
    abstract SubTextEquals: target: string * startIndex: int -> bool
    abstract Length: int
    abstract ContentEquals: sourceText: ISourceText -> bool
    abstract CopyTo: sourceIndex: int * destination: char[] * destinationIndex: int * count: int -> unit
```

With `getSubtext: (range -> string)`, a consumer (e.g., FSAC) that already had an `ISourceText` could just pass in its `GetSubTextFromRange` implementation directly:

```fsharp
let! unnecessaryParentheses =
    UnnecessaryParentheses.getUnnecessaryParentheses
        sourceText.GetSubTextFromRange
        parseResults.ParseTree
```

As for (2b), if we ever wanted to reap that benefit we'd either need to have `ISourceText.GetSubTextFromRange` return a `ReadOnlySpan<char>` instead of a `string` (even if the implementation still used `string`s under the hood for now), or else perhaps make a single-method interface like:

```fsharp
type IGetSubtext =
    abstract GetSubtext: range: range -> ReadOnlySpan<char>
```

A caller with an `ISourceText` could then do something like:

```fsharp
let getTextFromRange =
    { new IGetSubtext with
        member _.GetSubtext range =
            (sourceText.GetSubTextFromRange range).AsSpan() }

let! unnecessaryParentheses =
    UnnecessaryParentheses.getUnnecessaryParentheses
        getTextFromRange
        parseResults.ParseTree
```

…All of that to say: I vote that we either:

- Just use something like `getSubtext: (range -> string)` (or I guess maybe `getSourceLineStr: (int -> string)` if we prefer making it a bit easier on the caller), _or_
- Update #15979 to make `ISourceText.GetSubTextFromRange` return a `ReadOnlySpan<char>` and use that

[^1]: E.g.
    ```fsharp
    val getUnusedOpens:
        checkFileResults: FSharpCheckFileResults * getSourceLineStr: (int -> string) -> Async<range list>
    val getSimplifiableNames:
        checkFileResults: FSharpCheckFileResults * getSourceLineStr: (int -> string) -> Async<seq<SimplifiableRange>>
    ```

vzarytovskii commented on vsintegration/src/FSharp.Editor/Diagnostics/DocumentDiagnosticAnalyzer.fs:
```diff
@@ -108,7 +109,38 @@ type internal FSharpDocumentDiagnosticAnalyzer [<ImportingConstructor>] () =
-            if errors.Count = 0 then
+            let! unnecessaryParentheses =
+                match diagnosticType with
+                | DiagnosticsType.Semantic -> cancellableTask { return ImmutableArray.Empty }
```
This can just be a CancellableTask.singleton so we don't allocate and then run state machine.

brianrourkeboll replied:
Won't I eventually want to take this out of here anyway, though?

> I shoehorned the call to the analysis logic into `DocumentDiagnosticAnalyzer.fs` to test it out, but I assume we'd ultimately want to have a separate `UnnecessaryParenthesesDiagnosticAnalyzer.fs`.

> I assume I need to add `IFSharpUnnecessaryParenthesesDiagnosticAnalyzer` and `FSharpIDEDiagnosticIds.RemoveUnnecessaryParentheses` to Roslyn. Might that also be why the bulk-fix functionality is not working for this (because Roslyn is not aware that IDE0047 is supported for F#)?

vzarytovskii commented on vsintegration/src/FSharp.Editor/CodeFixes/RemoveUnnecessaryParentheses.fs:
```diff
@@ -0,0 +1,110 @@
+
+                match firstChar, lastChar with
+                | '(', ')' ->
+                    let inline toPat f x = if f x then Some() else None
+                    let (|LetterOrDigit|_|) = toPat Char.IsLetterOrDigit
+                    let (|Punctuation|_|) = toPat Char.IsPunctuation
+                    let (|Symbol|_|) = toPat Char.IsSymbol
```
These probably should be outside the state machine, so we're not codegen more than necessary

brianrourkeboll replied:
Hmm, I believe they're [compiled outside of the state machine](https://sharplab.io/#v2:DYLgZgzgPg9gDgUwHYAIDKBPCAXBBbAWAChjgFsUwUAKCFEFHAJwEskBzAShQF5iUBKbAEMIAaxQBvfoNlkKbYGwRCYABWEUqAD14oWVHUIAWydDDwrq3BMAgqAcjCQIZsgfJpQAMuVxMAeSYAERZ2FmwoAH0obh5VDQoAYWNhJgA6AEkIX2x/INDw7Dd3T2ooNQBXJABjbErNFmdo2L1sdU0UFLSsiCra+sbnErlyL0w8ACMYYBa4hM7ujOyJ6eARwQ2BPE0a40YAbQAGAF0UAHcI4y2UKBRc/JCwiJQAWgA+FCZySqZUAEYbnd+nUGtgmqgPl8fn8UAAmIHoDBTGZvT7feqwgDMiKiaOhmNQABYSgBfIA=) anyway. I could also just `inline` them.

brianrourkeboll replied:
aa04c18

vzarytovskii replied:
Yeah, generally rule of thumb - the less bindings you have inside the state machine instance - the less codegen, lifting and magic it needs to do to statically compile all the things.

vzarytovskii commented on src/Compiler/Service/ServiceAnalysis.fs:
```diff
@@ -451,3 +452,869 @@ module UnusedDeclarations =
+            open System
+
+            /// Tries to parse the given original notation as a symbolic infix operator.
+            let ofOriginalNotation (originalNotation: string) =
```
Can we return value option here instead?

brianrourkeboll replied:
Yep. I think I was originally calling this inside of `List.tryPick`, but I'm not anymore, so I'll update it.

vzarytovskii commented on src/Compiler/Service/ServiceAnalysis.fs:
```diff
@@ -451,3 +452,869 @@ module UnusedDeclarations =
+        /// Represents the infix application of a symbolic binary operator.
+        /// The original notation may include leading dots and trailing characters,
+        /// with the exception of TypeTest, Cons, Upcast, Downcast, BarBar, AmpAmp, and ColonEquals.
+        [<NoComparison; NoEquality>]
```
Can this be a struct?

brianrourkeboll replied:
aa04c18

brianrourkeboll commented:
...As for the logic itself: before anyone spends the time to do a thorough review, I should probably look over the ocamlformat implementation more closely for obvious things I'm missing (I threw most of this together before I knew it existed). There are of course a number of syntactic differences between OCaml and F# (and differences in how analogous constructs are represented in their respective ASTs), but most of the fundamentals are the same, and the ocamlformat implementation seems pretty battle-tested and likely covers many cases that I haven't thought of.

brianrourkeboll commented:
All right, I've made a few updates:

- I updated lambda traversal in `ServiceParseTreeWalk.fs` along the lines of the suggestion in https://github.com/dotnet/fsharp/pull/16079#discussion_r1348309389.
- I added handling for some more numeric literal cases.
- I added handling for expressions that would be parsed as type applications if the parens were removed (`(1<2),2>3`).
- I corrected some infix operator precedence assertions. Feel free to weigh in on whether you think the current balance is fine or whether we should be more conservative. The C# analyzer, for example, refuses to touch user-overloaded symbolic operators altogether, and it has special cases for when the arg type is a floating point type, etc.
- I'm now taking a somewhat more systematic approach to expression precedence, inspired in part by what ocamlformat does.[^1] It gives better coverage and cleans the code up a fair amount at the expense of some mental indirection in places.
- I replaced my first attempt at handling nested matches and if-then-elses with one that addresses other problematic dangling expressions as well. I don't _think_ it's worth adding memoization to it, but let me know if you think otherwise.
- I found that keeping the logic for handling directly-nested expressions and patterns separate from the more general logic for handling expressions and patterns in their respective AST contexts could be confusing and misleading, so I flattened them.

I've updated the main description with my remaining questions, but otherwise I think this PR might be ready for another look @vzarytovskii @nojaf (at your leisure).

[^1]: See, e.g.:
    - https://github.com/ocaml-ppx/ocamlformat/blob/d664fa2254479195f9fe0099c5847cb79299d17f/lib/Ast.ml#L1757-L1778
    - https://github.com/ocaml-ppx/ocamlformat/blob/d664fa2254479195f9fe0099c5847cb79299d17f/lib/Prec.ml#L13-L34
    - https://github.com/ocaml-ppx/ocamlformat/blob/d664fa2254479195f9fe0099c5847cb79299d17f/lib/Assoc.ml#L21-L37

vzarytovskii commented on vsintegration/src/FSharp.Editor/CodeFixes/RemoveUnnecessaryParentheses.fs:
```diff
@@ -13,6 +13,40 @@ open Microsoft.CodeAnalysis.Text
+                i <- i + 1
+
+        if found then i else -1
+#endif
```
Shall these be in the extensions module? We have some in Common, i believe.

brianrourkeboll replied:
Sure, I can move this to the `Microsoft.VisualStudio.FSharp.Editor.Extensions` module.

I have [some more](https://github.com/dotnet/fsharp/blob/b30c50349f2cb985dbf5b971770248a276431bc9/src/Compiler/Service/ServiceAnalysis.fs#L459-L503) in `ServiceAnalysis.fs`. Is there somewhere better in that project to put the latter?

vzarytovskii replied:
> I do have two questions about moving these extensions to a common location, though:
> 
> 1. Should they be made generic (`ReadOnlySpan<'T>`) instead of specific to `ReadOnlySpan<char>`?

I think depending on what we use now, if we only use `ROS<char>` let's include it

> 2. Depending on usage, the performance could differ significantly between these simple polyfills and the vectorized versions when running on .NET 7+. I don't think the perf difference should matter much in the common case for how they're used in this PR,[1](#user-content-fn-1-15a5e9655e909793370d8c9e06b9a8a2) but it's hard to say for other places they might be used once exposed in a central location.
> 

This particular piece of code is hosted in VS and currently will only run under net472 (i.e. ns20).

brianrourkeboll replied:
02edd60de5c3cfd017c005cecf4aa36cfe066f12

nojaf commented on src/Compiler/Service/ServiceAnalysis.fs:
```diff
@@ -451,3 +452,1068 @@ module UnusedDeclarations =
+
+                loop diveResults
+
+            let _ = SyntaxTraversal.traverseUntil pick parsedInput.Range.End visitor parsedInput
```
If this does what I think it does we should expose this publically in FCS.
My understanding is that this just keeps traversing and captures all the ranges in the HashSet?
@baronfel is this equivalent to what `SyntaxCollectorBase` does in FSAC?
This really might be worth it to expose as FCS API.

baronfel replied:
yeah, fair point - but that would require someone to write custom `pick`s. I much prefer the experience of having the two distinct modes:

* traverse the tree down to this point
* traverse the entire tree (within a given range)

brianrourkeboll replied:
> If this does what I think it does we should expose this publically in FCS. My understanding is that this just keeps traversing and captures all the ranges in the HashSet? @baronfel is this equivalent to what `SyntaxCollectorBase` does in FSAC? This really might be worth it to expose as FCS API.

Yes, the custom `pick` here (which never "picks" anything) lets it visit the whole AST. Some analyzers (where the goal is to generate diagnostics for a whole file) would indeed probably be much easier to implement if they had access to such a `traverseAll` function, which could be as simple as

```fsharp
let traverseAll (visitor: SyntaxVisitorBase<'T>) (parseTree: ParsedInput) : unit =
    let pick _ _ _ diveResults =
        let rec loop =
            function
            | [] -> None
            | (_, project) :: rest ->
                ignore (project ())
                loop rest

        loop diveResults

    ignore<'T option> (traverseUntil pick parseTree.Range.End visitor parseTree)
```

I can think of other possible signatures, though, e.g.,

```fsharp
val fold:
    folder: ('State -> SyntaxNode list -> SyntaxNode -> 'State) ->
    state: 'State
    tree: ParsedInput ->
        'State
```

where `folder` takes the state, path, and current node to produce the new state.

This can actually be implemented using the existing code like

```fsharp
let fold folder state tree =
    let mutable state = state

    let visitor =
        { new SyntaxVisitorBase<obj>() with
            member _.VisitExpr(path, _, defaultTraverse, expr) =
                state <- folder state path (SyntaxNode.SynExpr expr)
                defaultTraverse expr

            member _.VisitPat(path, defaultTraverse, pat) =
                state <- folder state path (SyntaxNode.SynPat pat)
                defaultTraverse pat

            // Override the rest likewise…
        }

    let pick _ _ _ diveResults =
        let rec loop =
            function
            | [] -> None
            | (_, project) :: rest ->
                ignore (project ())
                loop rest

        loop diveResults

    do ignore<obj option> (traverseUntil pick tree.Range.End visitor tree)
    state
```

It does make for pretty ergonomic usage, in my opinion:

```fsharp
let getUnnecessaryParentheses (getSourceLineStr: int -> string) (parsedInput: ParsedInput) : Async<range seq> =
    async {
        return upcast ((HashSet Range.comparer, parsedInput) ||> SyntaxTraversal.fold (fun acc path node ->
            match node with
            | SyntaxNode.SynExpr expr ->
                SynExpr.unnecessaryParentheses getSourceLineStr expr path
                |> ValueOption.map (fun range -> ignore (acc.Add range); acc)
                |> ValueOption.defaultValue acc
            | SyntaxNode.SynPat pat ->
                SynPat.unnecessaryParentheses pat path
                |> ValueOption.map (fun range -> ignore (acc.Add range); acc)
                |> ValueOption.defaultValue acc
            | _ -> acc))
    }
```

You could also have something like this:

```fsharp
val foldWhile:
    folder: ('State -> SyntaxNode list -> SyntaxNode -> (SyntaxNode * 'State) option) ->
    state: 'State
    tree: SyntaxNode ->
        'State
```

...where the folder returns which subnode to dive into next as well as the new state, which could also be used to implement the existing `Traverse`.

nojaf replied:
This looks interesting. I think we should indeed not expose FCS consumers to write a custom `pick` function. But indeed exposing a `traverse until point` and `traverse entire tree` would be quite valuable.

I'm a little less convinced of the `folder: ('State -> SyntaxNode list -> SyntaxNode -> 'State)` idea. It looks nice but feels like a departure from the `SyntaxVisitorBase`. You would also ignore most `SyntaxNode` in a pattern match if you use this. While the `SyntaxVisitorBase` keeps it somewhat explicit what types of nodes you are hooking into.

We can circle back to this in a future PR.

brianrourkeboll replied:
I ended up exposing

```fsharp
val internal traverseAll: visitor: SyntaxVisitorBase<'T> -> parseTree: ParsedInput -> unit
```

internally for now in 19da092 to make the call inside of `getUnnecessaryParentheses` clearer. `traverseUntil` is still there (`traverseAll` and `Traverse` both call it), but it's private to the `SyntaxTraversal` module.

psfinaki replied:
Resolving this since this can also be done in a followup.

nojaf reviewed (approved):
Looks good from my end. I have not reviewed anything in the `vsintegration` folder.
Please [format your code](https://github.com/dotnet/fsharp/blob/main/DEVGUIDE.md#automated-source-code-formatting) once you are done.

Great work!

nojaf commented:
I think you also need another update to the Surface area.

I always run

```pwsh
$env:TEST_UPDATE_BSL=1 
dotnet test tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.Tests.fsproj --filter "SurfaceAreaTest" /p:BUILDING_USING_DOTNET=true
dotnet test tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.Tests.fsproj --filter "SurfaceAreaTest" /p:BUILDING_USING_DOTNET=true
dotnet test tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.Tests.fsproj --filter "SurfaceAreaTest" -c Release /p:BUILDING_USING_DOTNET=true
dotnet test tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.Tests.fsproj --filter "SurfaceAreaTest" -c Release /p:BUILDING_USING_DOTNET=true
```

and get it over with.

This updates 
![image](https://github.com/dotnet/fsharp/assets/2621499/fe114b88-99da-4ab2-9ff8-67f453ca755e)

brianrourkeboll commented:
@vzarytovskii I made some updates under the assumption that we _wouldn't_ be trying to make any upstream changes in Roslyn.

That does mean that the unnecessary parens analysis will run in series in the call to `DocumentDiagnosticAnalyzer.AnalyzeSyntaxAsync` (via `DocumentDiagnosticAnalyzer.GetDiagnostics`) with the analysis done in `GetFSharpParseResultsAsync` rather than concurrently, such that the task returned by `AnalyzeSyntaxAsync` that holds the parsing diagnostics must await the completion of `UnnecessaryParentheses.getUnnecessaryParentheses`. This should only add the cost of an untyped AST traversal as long as there are no perf holes in my code (or in `SyntaxTraversal` itself).

As far as I can tell, I'm not sure there's a way around that without adding `IFSharpUnnecessaryParenthesesDiagnosticAnalyzer`, `FSharpIDEDiagnosticIds.RemoveUnnecessaryParentheses`, etc., to Roslyn so that we can export a separate analyzer here. But that's just my guess—maybe there is another way that I just don't see.

brianrourkeboll commented:
I wonder whether this code in Roslyn might be the reason the fix-all functionality does not work for IDE0047:

https://github.com/dotnet/roslyn/blob/b2aa42e7458f5ca611124a14b3552aa9e2cbd7ba/src/Tools/ExternalAccess/FSharp/Internal/Diagnostics/FSharpDocumentDiagnosticAnalyzer.cs#L56-L65

#### Fix all not working for IDE0047
![image](https://github.com/dotnet/fsharp/assets/14795984/fddf9349-e178-4659-842f-928eecf90b34)

If I use an arbitrary new FSxxxx diagnostic ID instead of IDE0047, the fix-all functionality works as expected:

#### Fix all working for FSxxxx
![image](https://github.com/dotnet/fsharp/assets/14795984/17f8f876-be67-45f7-8d1d-59892567f70a)

That's rather unfortunate. What's the best course of action here?

1. Just use a new FSxxxx diagnostic ID instead of IDE0047?
2. Just use a new FSxxxx diagnostic ID instead of IDE0047 so that we don't need to wait for Roslyn, but also start the process of updating Roslyn so that we can eventually use IDE0047?
3. Start the process of updating Roslyn and just accept that it may take a while for it to make it back here? (The underlying `getUnneededParentheses` API could still be merged so that non-VS tooling could use it.)

Perhaps one of the main reasons to do (2) or (3) would be if we ever planned to support any of the existing IDE0047-related .editorconfig options called out in #16078, although we could in theory support them separately for a new FSxxxx ID as well. I guess another one would be if we ever wanted to fully support _any_ non-FSxxxx diagnostic IDs in the VS tooling.

T-Gro commented:
> I wonder whether this code in Roslyn might be the reason the fix-all functionality does not work for IDE0047:

Yes, that is likely the reason.
Roslyn has to be aware of the domain of possible error codes supported by the F# plugin.

psfinaki commented on vsintegration/src/FSharp.Editor/Common/Extensions.fs:
```diff
@@ -572,3 +572,24 @@ type Async with
+[<Sealed; AbstractClass; Extension>]
+type ReadOnlySpanExtensions =
+    [<Extension>]
+    static member IndexOfAnyExcept(span: ReadOnlySpan<char>, value0: char, value1: char) =
```
Honestly it took me a while to get point here :) I think I was confused by the name since `IndexOfAny` has a bit different idea. Not that I can come up with a better name yet though. `IndexOfOther` probably wouldn't help much.

brianrourkeboll replied:
Yeah, this and the ones in `ServiceAnalysis.fs` are just (non-vectorized) polyfills for the [APIs added to `System.MemoryExtensions` in .NET 7](https://learn.microsoft.com/en-us/dotnet/api/system.memoryextensions.indexofanyexcept?view=net-7.0). (The names were landed on here: https://github.com/dotnet/runtime/issues/28795#issuecomment-1097040761.)

psfinaki replied:
Ohh okay then, let's keep the name.

brianrourkeboll replied:
The name notwithstanding, they are super handy methods :)

psfinaki reviewed (commented):
@brianrourkeboll hey, this is a phenomenal piece of work - thanks for this. It would be a great addition to our analyzers. I left a couple of notes and will do another review soon.

Regarding Roslyn - I'd vote for the following: let's not involve Roslyn here and change the diagnostic number for now. We can reimplement things through Roslyn as a followup. 

Involving Roslyn would take time (weeks to months) whereas this can be useful right now. Moreover, we want to first do some strategy planning about general interaction with Roslyn in respect to LSP, analyzers and so on. 

And in any case, we will be able to first update the Roslyn in the way we want and then just move to the new interface/existing diag number without the functionality disruption.

psfinaki commented on vsintegration/src/FSharp.Editor/CodeFixes/RemoveUnnecessaryParentheses.fs:
```diff
@@ -0,0 +1,165 @@
+
+    /// IDE0047: Remove unnecessary parentheses.
+    ///
+    /// https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0047-ide0048
```
No, not necessarily. We have a doc about code fixes with some links, so you can remove this.

brianrourkeboll commented on vsintegration/src/FSharp.Editor/CodeFixes/RemoveUnnecessaryParentheses.fs:
```diff
@@ -0,0 +1,165 @@
+            assert (context.Span.Length >= 3) // (…)
+
+            cancellableTask {
+                let! sourceText = context.Document.GetTextAsync context.CancellationToken
```
caccbf9

psfinaki commented on src/Compiler/Service/ServiceAnalysis.fsi:
```diff
@@ -31,3 +32,12 @@ module public UnusedDeclarations =
+    ///
+    /// Note that this may include pairs of nested ranges each of whose
+    /// lack of necessity depends on the other's presence, such
+    /// that it is valid to remove either set of parentheses but not both.
```
Do you think some tests could be also added on the compiler service level?

Also, it would be helpful to specify some examples in the docs here. E.g. I logically understand the above sentence but it took me a while to come up with an example.

brianrourkeboll replied:
Yes, I'll do both.

brianrourkeboll replied:
a7cca0f

brianrourkeboll replied:
I added some very basic tests that directly call `FSharp.Compiler.EditorServices.UnnecessaryParentheses.getUnnecessaryParentheses` here:

a2b70f9

Is that what you were thinking? Or should they be more comprehensive, like the code fix tests?

Also, `getUnnecessaryParentheses` will actually _not_ currently return two ranges for the example that I added to the doc comment here in a7cca0f (`(x.M(y)).N`), even though it theoretically could (and did at some point).

One possibility might simply be to update `getUnnecessaryParentheses` so that it has an opinion on all known mutually-dependent nested parentheses scenarios (`(x.M(y)).N`, `id (-(-x))`, …?), instead of filtering them out at the fix-all stage. The only problem with that approach is that I don't know how we can be 100% sure that we've thought of all such possible scenarios.

If we could somehow be confident in that, or confident enough, then that would be nicer, since as it is you must explicitly apply successive rounds of the code fix to strip nested unnecessary parentheses: `((3))` → `(3)` → `3`, `((x + y) + z)` → `(x + y) + z` → `x + y + z`, etc.

psfinaki replied:
I think the new tests are enough. Since this is a complex matter, I am quite sure we might have not taken some corner cases into account - but that's fine, that's our approach, IMO as long as there are easy ways to repro and test things, we can quickly iterate and followup.

For about the same reason, I am also inclined to have this opinionated decisions about mutually dependent nested parentheses. And well, since this is a minority of cases, we can just pick the first possible option and see this in action.

brianrourkeboll replied:
Hmm. It looks like even if we did take a stance on `(x.M(y)).N → (x.M y).N ↮ x.M(y).N`, etc., the way the fixes are currently applied during a fix-all application doesn't make sense when there are nested fixes, even when they are _not_ mutually-dependent. That's because the ranges to which any inner fixes would apply can be shifted somewhat unpredictably by the application of outer fixes: e.g., when an open paren is removed, a leading space may or may not be inserted to replace it depending on the context, etc.

The C#/VB code fix _does_ appear to handle this, though I have not yet looked at how it does so.

brianrourkeboll replied:
...It looks like the C#/VB fix-all implementation has access to some more sophisticated AST-editing machinery than we currently do:

https://github.com/dotnet/roslyn/blob/79d1884dcdc0fb3f3c0763e372e2e010dbbcfa4a/src/Analyzers/Core/CodeFixes/RemoveUnnecessaryParentheses/AbstractRemoveUnnecessaryParenthesesCodeFixProvider.cs#L34-L47

I can think of a few things to experiment with, but I won't have time till next week, so maybe it can be a followup.

psfinaki replied:
Yeah I think we can take inspiration from there - thanks for looking into this.

psfinaki commented on src/Compiler/Service/ServiceAnalysis.fs:
```diff
@@ -451,3 +452,1091 @@ module UnusedDeclarations =
+
+#if !NET7_0_OR_GREATER
+    [<Sealed; AbstractClass; Extension>]
+    type ReadOnlySpanExtensions =
```
Feel free to split those to a separate module/file.

brianrourkeboll replied:
c96ec3a

psfinaki commented on src/Compiler/Service/ServiceAnalysis.fs:
```diff
@@ -451,3 +452,1091 @@ module UnusedDeclarations =
+            | Set, Set -> 0
+
+    /// Associativity/association.
+    type Assoc =
```
BTW I cannot help having the IngSoc association here, partially because you have 2 + 2 = 5 in the PR description :D

psfinaki reviewed (commented):
A bunch of small nitpicks to the service analysis - feel free to ignore :)

psfinaki commented on src/Compiler/Service/ServiceParseTreeWalk.fs:
I am alright with the changes here for now - although overall I feel we're missing some holistic approach in how and where we expose data from the compiler to VS. It looks like sometimes we randomly add things to `ServiceParseTreeWalk`, other times to `FSharpParseFileResults`, yet other times we process parse results ad-hoc in the editor layer.

If you have any thoughts on this - now or whenever - feel free to share, some fresh perspective would be very useful here.

psfinaki commented on vsintegration/tests/FSharp.Editor.Tests/CodeFixes/RemoveUnnecessaryParenthesesTests.fs:
That's truly amazing testing.

A small note on the file structure - it would be awesome to keep the framework bits in the `CodeFixTestFramework` - recently we've tried to clean things up here and any step aside might activate the broken windows theory :) It would be great to really have just "broken code -> fixed code" bits here, thereby these test modules have some documentation function also.

I would probably also split one-line and multi-line cases to different test modules - again, just to keep things more readable in that regard.

brianrourkeboll replied:
Thanks for the kind words!

> it would be awesome to keep the framework bits in the `CodeFixTestFramework`

Yes, I can look to see what in here can be hoisted up into the shared code.

> It would be great to really have just "broken code -> fixed code" bits here

I'm not completely sure what you mean by this. Do you mean separating "expect fix" examples from "expect no fix" examples? That's what I started out doing, but it became really hard to keep track of minimal pairs that were _almost_ identical but where a fix should apply for one and not the other. By having them interleaved, you can have such minimal pairs right next to each other.

> I would probably also split one-line and multi-line cases to different test modules - again, just to keep things more readable in that regard.

Yes, that would probably help with readability in one way, but I'm afraid that we'd again lose the advantage of proximity to similar examples, e.g., it's clear here that these are both good and related tests of how lambdas to the right of an infix operator are handled, whereas if the multiline example were somewhere else, it would take more effort to find and compare them:

```fsharp
"x |> (fun x -> x)", "x |> fun x -> x"

"
x
|> id
|> id
|> id
|> id
|> id
|> (fun x -> x)
",
"
x
|> id
|> id
|> id
|> id
|> id
|> fun x -> x
"
```

I guess I could replace all the comments I have in there with modules, and each module could have single-line and multiline submodules, which would help maintain the advantages of proximity:

```fsharp
module FunctionApplications =
    module Lambdas =
        module SingleLine = // …
        module Multiline = // …
```

But I was doing some more granular breakdowns earlier on, and the deeply nested nodes got pretty unwieldy in the test explorer pretty fast (e.g., I had to click a lot of expansion arrows to drill down to a single failed test).

I'll do a pass and see what kind of cleanup I can do.

psfinaki replied:
> I'm not completely sure what you mean by this. 

Nah that was just a repetition of the previous thought :) Framework bits in the framework part, test bits in tests. 

> I'll do a pass and see what kind of cleanup I can do.

Your arguments make sense, proximity is important here. I'll leave it up to you - modules might or might make things clearer.

brianrourkeboll replied:
@psfinaki

I moved some framework-y code from my tests up into the test framework file in dfdcab3. Let me know whether you think what I did makes sense.

I had to refactor a couple things in there a tiny bit to support both the existing tests and the new tests for this code fix, since these new ones rely on a call to `FSharpDocumentDiagnosticAnalyzer.GetDiagnostics` instead of just being able to use the diagnostics that are emitted as part of parsing and checking.

I may or may not have time to do some more cleanup tomorrow, but then I probably won't have time to look again for about a week, so you're free to leave this alone till then or merge it as-is.

psfinaki replied:
This looks good to me. Let me take another fresher look tomorrow morning - nevertheless unless we discover some bigger gaps in the implementation or testing, I am inclined to let this in now and do some followups when we have capacity for it :)

psfinaki commented:
@brianrourkeboll I have also now thought about cases of broken code like `let test = (2 + (3 + 4)`. Does this analyzer bother with those currently? 

I don't have a strong opinion on if it should or should not do something about this. And unfortunately we don't have a proper testing framework for analyzers yet to autotest this. So if you share your perspective and add these cases to the demo, it would be helpful.

brianrourkeboll commented on vsintegration/src/FSharp.Editor/CodeFixes/RemoveUnnecessaryParentheses.fs:
```diff
@@ -0,0 +1,165 @@
+
+    override this.GetFixAllProvider() =
+        this.RegisterFsharpFixAll(fun diagnostics ->
+            // There may be pairs of diagnostics with nested spans
```
Yeah, any given diagnostic that hits `GetCodeFixIfAppliesAsync` will only be for a single pair of parens, but in the fix-all scenario, you could have code like the following where the analyzer will generate two diagnostics—one for the inner pair and one for the outer, each of which is valid on its own—but where removing both pairs at the same time would be invalid:

```fsharp
let (~-) … = …

id (-(-x))
id -(-x)        // Valid.
id (- -x)       // Also valid.
id - -x         // Not valid.

id (-(<@ x @>))
id -(<@ x @>)   // Valid.
id (- <@ x @>)  // Valid.
id - <@ x @>    // Not valid.

id (-($"…"))
id -($"…")      // Valid
id (- $"…")     // Valid
id - $"…"       // Not valid.
```

Those are maybe not the most compelling examples, but they are possible, and it seems like there could be others.

psfinaki replied:
Ufff yeah, you're right. That's a valid case and this code should be here then. Good to know that such things can happen.

brianrourkeboll commented on src/Compiler/Service/ServiceAnalysis.fs:
```diff
@@ -451,3 +452,1091 @@ module UnusedDeclarations =
+    /// Represents the precedence of a binary-ish expression.
+    type Precedence =
+        /// <-
+        | Set
```
I think I probably based this name on these:

https://github.com/dotnet/fsharp/blob/f88d5d49f617589f4a77a8f4d51ad5c68d943ce6/src/Compiler/SyntaxTree/SyntaxTree.fsi#L797-L798
https://github.com/dotnet/fsharp/blob/f88d5d49f617589f4a77a8f4d51ad5c68d943ce6/src/Compiler/SyntaxTree/SyntaxTree.fsi#L794-L795
https://github.com/dotnet/fsharp/blob/f88d5d49f617589f4a77a8f4d51ad5c68d943ce6/src/Compiler/SyntaxTree/SyntaxTree.fsi#L785-L786
https://github.com/dotnet/fsharp/blob/f88d5d49f617589f4a77a8f4d51ad5c68d943ce6/src/Compiler/SyntaxTree/SyntaxTree.fsi#L803-L821

brianrourkeboll commented on src/Compiler/Service/ServiceAnalysis.fs:
```diff
@@ -451,3 +452,1091 @@ module UnusedDeclarations =
+            | Exp, _ -> 1
+            | _, Exp -> -1
+
+            | (Mod | Div | Mul), (Mod | Div | Mul) -> 0
```
Hmm, I do need some way to distinguish them here:

https://github.com/dotnet/fsharp/blob/d2f1666a8e5bebae21ee15cda96e284a74cf0597/src/Compiler/Service/ServiceAnalysis.fs#L1264-L1269

E.g., `x * (y * z)` → `x * y * z` is fine, but `x * (y / z)` → `x * y / z` is not.

We could alternatively keep precedence level and expression kind separate, even though we don't always need to inspect the latter.

psfinaki replied:
I would personally find having those things separate a bit more clear but it's definitely not a big deal, feel free not to bother.

psfinaki replied:
Thanks for giving this a try. Well, one downside of adding more constructs here is that each needs a name which is the hardest thing in life. So I think we can suffice with comments for now and we can come back to this later if needed.

brianrourkeboll replied:
I added a couple comments in 19637f3. I do agree that having `Precedence` be dual-purpose like it is right now is not the cleanest.

brianrourkeboll commented:
@psfinaki

> @brianrourkeboll I have also now thought about cases of broken code like `let test = (2 + (3 + 4)`. Does this analyzer bother with those currently?
> 
> I don't have a strong opinion on if it should or should not do something about this. And unfortunately we don't have a proper testing framework for analyzers yet to autotest this. So if you share your perspective and add these cases to the demo, it would be helpful.

It currently only returns ranges for complete pairs of parens—note the match on `rightParenRange = Some _` here:

https://github.com/dotnet/fsharp/blob/d2f1666a8e5bebae21ee15cda96e284a74cf0597/src/Compiler/Service/ServiceAnalysis.fs#L1203-L1204

The code fix likewise operates only on pairs:

https://github.com/dotnet/fsharp/blob/d2f1666a8e5bebae21ee15cda96e284a74cf0597/vsintegration/src/FSharp.Editor/CodeFixes/RemoveUnnecessaryParentheses.fs#L111

https://github.com/dotnet/fsharp/blob/d2f1666a8e5bebae21ee15cda96e284a74cf0597/vsintegration/src/FSharp.Editor/CodeFixes/RemoveUnnecessaryParentheses.fs#L120-L121

So, in your example, `let test = (2 + (3 + 4)`, the analyzer will suggest `let test = (2 + 3 + 4`, but it won't suggest removing the extra open paren.

I guess that supporting that, if we wanted to, would be as easy as categorically returning the open paren range when there is no closing paren. We could even to offer to insert a closing paren, though that would be more involved…

I notice that the C# code fix in VS behaves the same way in this case as this one currently does, for what that's worth.

Either way, I agree with you that it's a good idea to add some tests for this. Thanks!

brianrourkeboll commented on vsintegration/src/FSharp.Editor/CodeFixes/RemoveUnnecessaryParentheses.fs:
```diff
@@ -0,0 +1,165 @@
+        let startCol = startLinePosition.Character
+        let endLine = endLinePosition.Line
+
+        if startLine = endLine then
```
Hmm, there are indeed probably other scenarios where it'd be useful to be aware of potential offside-line skewing, but this code is currently at least partly parentheses-specific, and for other uses you'd probably need to make it agnostic of open/close tokens, open/close token length, etc.: for example, maybe you couldn't just use `IndexOfAnyExcept(' ', '(')`, etc., since the open or close token is a full word or whatever, and so on. I think I'll leave that for another day :)

psfinaki replied:
Alrigjhty, let's keep it here and split it if/once necessary.

brianrourkeboll commented on src/Compiler/Service/ServiceAnalysis.fs:
```diff
@@ -451,3 +452,1091 @@ module UnusedDeclarations =
+            if found then i else -1
+#endif
+
+    /// Represents the precedence of a binary-ish expression.
```
addc8d6

psfinaki commented:
/azp run

psfinaki commented:
> I notice that the C# code fix in VS behaves the same way in this case as this one currently does, for what that's worth.

That's a good bonus then. I think we can later consider adding a "remove the unmatched parenthesis" code fix, although I wonder why C# didn't introduce it yet.

psfinaki commented on vsintegration/src/FSharp.Editor/CodeFixes/CodeFixHelpers.fs:
```diff
@@ -194,3 +192,9 @@ module IFSharpCodeFixProviderExtensions =
                 provider.FixAllAsync fixAllCtx doc allDiagnostics
                 |> CancellableTask.start fixAllCtx.CancellationToken)
+
+        member provider.RegisterFsharpFixAll filter =
```
Maybe you can then rewrite the existing overload via this new one, as it's more generic now. Or feel free not to bother and resolve.

brianrourkeboll replied:
Yeah I think having the two overloads is fine; I was just referring to the lowercase `s`

```diff
- RegisterFsharpFixAll
+ RegisterFSharpFixAll
```

:)

psfinaki commented on vsintegration/tests/FSharp.Editor.Tests/CodeFixes/CodeFixTestFramework.fs:
```diff
@@ -17,118 +17,273 @@ open FSharp.Editor.Tests.Helpers
+    |> CancellableTask.runSynchronouslyWithoutCancellation
+
+/// Contains types and functions for coveniently making code fix assertions using xUnit.
+[<AutoOpen>]
```
This should probably go to the `FSharp.Test.Utilities` project, the module `Assert.fs` is there basically for these purposes.

brianrourkeboll replied:
Hmm. They could, although they're somewhat tailored to code fix tests specifically (the `MissingCodeFixException` versus `UnexpectedCodeFixException` versus `WrongCodeFixException`, etc.).

The `memberData` CE _could_ be more broadly useful, although other uses might want additional overloads (taking triples instead of pairs), etc.

psfinaki replied:
Sure, we can move this once there is actually a need for reusing this code.

brianrourkeboll commented on vsintegration/tests/FSharp.Editor.Tests/CodeFixes/PrefixUnusedValueTests.fs:
```diff
@@ -57,25 +57,4 @@ let f() =
 
     Assert.Equal(expected, actual)
 
-[<Fact>]
```
See 13c28405020d8dad69b231afc3526a5635901e70. That's because the diagnostics are now being filtered down to fixable ones, although that can be reverted if needed.

brianrourkeboll commented on vsintegration/tests/FSharp.Editor.Tests/CodeFixes/RemoveUnnecessaryParenthesesTests.fs:
```diff
@@ -0,0 +1,1573 @@
+[<AutoOpen>]
+module private TopLevel =
+    let private fixer = FSharpRemoveUnnecessaryParenthesesCodeFixProvider()
+    let private fixAllProvider = fixer.RegisterFsharpFixAll()
```
Whoops, looks like I accidentally kept that from my bulk fix test experimentation…

brianrourkeboll commented:
> @brianrourkeboll Left a few comments with small ideas for cleanup - but generally this LGTM, great job!
> 
> I guess that anyway the whole thing will benefit from a fresh look. So if you think you won't get to fix these small things today, I can resolve the comments and you can address them later as well.

Some good catches, but yeah, I don't have my computer with me, so I probably won't be able to address them till next week.

psfinaki commented:
So @brianrourkeboll I am setting this one to merge - let's have another take on this next week :) Meanwhile I am looking forward to give it a try and apply to our code base!

brianrourkeboll commented on vsintegration/tests/FSharp.Editor.Tests/CodeFixes/CodeFixTestFramework.fs:
```diff
@@ -17,118 +17,273 @@ open FSharp.Editor.Tests.Helpers
-    cancellableTask {
-        let! _, checkFileResults = document.GetFSharpParseAndCheckResultsAsync "test"
+module ValueOption =
+    let inline toOption o =
```
I did try that, but it didn't work because they're marked `inline` (even though FSharp.Editor has its internals visible to FSharp.Editor.Tests):

```
1>D:\src\fsharp\vsintegration\src\FSharp.Editor\Common\Extensions.fs(325,16): error FS1113: The value 'ofOption' was marked inline but its implementation makes use of an internal or private function which is not sufficiently accessible
1>D:\src\fsharp\vsintegration\tests\FSharp.Editor.Tests\CodeFixes\CodeFixTestFramework.fs(89,12): warning FS1116: A value marked as 'inline' has an unexpected value
1>D:\src\fsharp\vsintegration\tests\FSharp.Editor.Tests\CodeFixes\CodeFixTestFramework.fs(89,12): error FS1118: Failed to inline the value 'ofOption' marked 'inline', perhaps because a recursive value was marked 'inline'
1>Done building project "FSharp.Editor.Tests.fsproj" -- FAILED.
```

It feels like `toOption` and `ofOption` should really just be on FSharp.Core's `ValueOption` module anyway :)

brianrourkeboll commented on src/Compiler/Service/ServiceAnalysis.fs:
```diff
@@ -451,3 +452,1056 @@ module UnusedDeclarations =
+        async {
+            let ranges = HashSet Range.comparer
+
+            let visitor =
```
@psfinaki Hmm, if you're referring specifically to the use of the visitor, then `FSharpParseFileResults` seems like it would be a better fit than `ServiceParseTreeWalk`, although I'd still lean towards how it is here. I was basically following the model of `UnusedOpens.getUnusedOpens`, `SimplifyNames.getSimplifiableNames`, and `UnusedDeclarations.getUnusedDeclarations` in this file, except that those use the check results instead of the parse results.

These modules and their single public functions in `FSharp.Compiler.EditorServices` do seem like they could perhaps be reorganized under modules dedicated to the AST (`FSharpParseFileResults`) and TAST (`FSharpCheckFileResults`), and then either applied to `FSharpParseFileResults` or `FSharpCheckFileResults` values or else exposed as instance methods on those types.

psfinaki replied:
Yeah it is part of a bigger problem in a way, we have similar pieces of code all over the place without much guidelines. So if you have an idea how to organize this properly, your fresh view would be helpful.
